### PR TITLE
Add TCP Reuse Port and TLS1.3 Ciphersuites Customize Feature

### PR DESCRIPTION
**Pull request by xsm1997**
1. Add TCP reuse port feature. This allow user to start multiple instance of Trojan and should improve TCP throughput on multi-core systems.
2. Add TLS1.3 ciphersuites customize feature. Trojan used to only be able to customize TLS1.2 ciphers. This feature allow user to customize TLS1.3 ciphersuite with OpenSSL >= 1.1.1 compiled.

### CODE REVIEW

**Comment by GreaterFire** [comment]
@xsm1997 I'll get back to you late December due to IRL responsibilities. But TCP port reuse is by default enabled just a few lines above your code.

**Comment by GreaterFire** [comment]
@xsm1997 Also, I believe that you can set TLS 1.3 ciphersuite with SSL_CTX_set_cipher_list

**Comment by xsm1997** [comment]
> @xsm1997 I'll get back to you late December due to IRL responsibilities. But TCP port reuse is by default enabled just a few lines above your code.

You enabled TCP address reuse, not port reuse, as far as I see.

**Comment by xsm1997** [comment]
> @xsm1997 Also, I believe that you can set TLS 1.3 ciphersuite with SSL_CTX_set_cipher_list

Unfortunately, No. You cannot set TLS 1.3 ciphersuite using SSL_CTX_set_cipher_list. It is a different API by OpenSSL documentation. [(Openssl Documentation)](https://www.openssl.org/docs/manmaster/man3/SSL_CTX_set_ciphersuites.html)

**Comment by zhangsan946** [comment]
@GreaterFire 大火兄，请review一下pr，merge一下撒~

File: docs/config.md [hunk 1]
```diff
@@ -34,6 +35,7 @@ In this page, we will look at the config file of trojan. Trojan uses [`JSON`](ht
     "tcp": {
         "no_delay": true,
         "keep_alive": true,
+        "reuse_port": false,
```
Comment (GreaterFire) [inline docs/config.md:38]: Add this also to other config types.

File: src/core/config.h [hunk 2]
```diff
@@ -67,6 +68,7 @@ class Config {
         bool keep_alive;
         bool fast_open;
         int fast_open_qlen;
+        bool reuse_port;
```
Comment (GreaterFire) [inline src/core/config.h]: Move it up between `keep_alive` and `fast_open`.

File: src/core/config.cpp [hunk 3]
```diff
@@ -90,6 +91,7 @@ void Config::populate(const ptree &tree) {
     tcp.keep_alive = tree.get("tcp.keep_alive", true);
     tcp.fast_open = tree.get("tcp.fast_open", false);
     tcp.fast_open_qlen = tree.get("tcp.fast_open_qlen", 20);
+    tcp.reuse_port = tree.get("tcp.reuse_port", false);
```
Comment (GreaterFire) [inline src/core/config.cpp]: Move it up between keep_alive and fast_open.

File: src/core/service.cpp [hunk 4]
```diff
@@ -37,6 +37,10 @@ using namespace std;
 using namespace boost::asio::ip;
 using namespace boost::asio::ssl;
 
+#ifdef ENABLE_REUSE_PORT
+typedef boost::asio::detail::socket_option::boolean<SOL_SOCKET, SO_REUSEPORT> reuse_port;
+#endif
```
Comment (GreaterFire) [inline src/core/service.cpp]:
```suggestion
#endif // ENABLE_REUSE_PORT
```

File: src/core/service.cpp [hunk 5]
```diff
@@ -53,6 +57,17 @@ Service::Service(Config &config, bool test) :
         tcp::endpoint listen_endpoint = *resolver.resolve(config.local_addr, to_string(config.local_port)).begin();
         socket_acceptor.open(listen_endpoint.protocol());
         socket_acceptor.set_option(tcp::acceptor::reuse_address(true));
+
+#ifdef ENABLE_REUSE_PORT
+        if (config.tcp.reuse_port) {
+            socket_acceptor.set_option(reuse_port(true));
+        }
+#else
```
Comment (GreaterFire) [inline src/core/service.cpp]:
```suggestion
#else // ENABLE_REUSE_PORT
```

File: src/core/service.cpp [hunk 6]
```diff
@@ -53,6 +57,17 @@ Service::Service(Config &config, bool test) :
...
+        if (config.tcp.reuse_port) {
+            Log::log_with_date_time("TCP_REUSEPORT is not supported", Log::WARN);
+        }
+#endif
```
Comment (GreaterFire) [inline src/core/service.cpp]:
```suggestion
#endif // ENABLE_REUSE_PORT
```

File: src/core/service.cpp [hunk 7]
```diff
@@ -168,6 +183,15 @@ Service::Service(Config &config, bool test) :
     if (config.ssl.cipher != "") {
         SSL_CTX_set_cipher_list(native_context, config.ssl.cipher.c_str());
     }
+#ifdef ENABLE_TLS13_CIPHERSUITES
+    if (config.ssl.cipher_tls13 != "") {
+        SSL_CTX_set_ciphersuites(native_context, config.ssl.cipher_tls13.c_str());
+    }
+#else
+    if (config.ssl.cipher_tls13 != "") {
+        Log::log_with_date_time("TLS1.3 ciphersuites is not supported", Log::WARN);
```
Comment (GreaterFire) [inline src/core/service.cpp]:
```suggestion
        Log::log_with_date_time("TLS1.3 ciphersuites are not supported", Log::WARN);
```

**Review by GreaterFire — CHANGES_REQUESTED** [review]
Code looks very good to me! In addition to the comments, can you also add these 2 config options to examples and tests?
Once you sign the CLA I'll merge it.

File: src/core/service.cpp [hunk 8]
```diff
@@ -168,6 +183,15 @@ Service::Service(Config &config, bool test) :
...
     }
+#ifdef ENABLE_TLS13_CIPHERSUITES
+    if (config.ssl.cipher_tls13 != "") {
+        SSL_CTX_set_ciphersuites(native_context, config.ssl.cipher_tls13.c_str());
+    }
+#else
+    if (config.ssl.cipher_tls13 != "") {
+        Log::log_with_date_time("TLS1.3 ciphersuites are not supported", Log::WARN);
+    }
+#endif
```
Comment (wongsyrone) [inline src/core/service.cpp]: I suggest

```cpp
    if (config.ssl.cipher_tls13 != "") {
#ifdef ENABLE_TLS13_CIPHERSUITES
        SSL_CTX_set_ciphersuites(native_context, config.ssl.cipher_tls13.c_str());
#else
        Log::log_with_date_time("TLS1.3 ciphersuites are not supported", Log::WARN);
#endif
    }
```

Comment (wongsyrone) [reply src/core/service.cpp]: BTW, you'd better add comments for these preprocessor if directives
```cpp
    if (config.ssl.cipher_tls13 != "") {
#ifdef ENABLE_TLS13_CIPHERSUITES
        SSL_CTX_set_ciphersuites(native_context, config.ssl.cipher_tls13.c_str());
#else  // ENABLE_TLS13_CIPHERSUITES
        Log::log_with_date_time("TLS1.3 ciphersuites are not supported", Log::WARN);
#endif // ENABLE_TLS13_CIPHERSUITES
    }
```

File: src/core/service.cpp [hunk 9]
```diff
@@ -53,6 +57,17 @@ Service::Service(Config &config, bool test) :
...
+
+#ifdef ENABLE_REUSE_PORT
+        if (config.tcp.reuse_port) {
+            socket_acceptor.set_option(reuse_port(true));
+        }
+#else // ENABLE_REUSE_PORT
+        if (config.tcp.reuse_port) {
+            Log::log_with_date_time("TCP_REUSEPORT is not supported", Log::WARN);
+        }
+#endif // ENABLE_REUSE_PORT
```
Comment (wongsyrone) [inline src/core/service.cpp:66]: Please fold preprocessor if directives into the if clause as cipher_tls13.

**Comment by wongsyrone** [comment]
Can you add <SOL_SOCKET, SO_REUSEADDR> , it's pretty like SO_REUSEPORT, just put the option above the current so_reuseport.

http://man7.org/linux/man-pages/man7/socket.7.html

**Comment by xsm1997** [comment]
@wongsyrone SO_REUSEADDR is already enabled a few lines above using the boost built-in way.

**Review by wongsyrone — APPROVED** [review]
LGTM now.

File: CMakeLists.txt [hunk 10]
```diff
@@ -71,6 +76,11 @@ if(CMAKE_SYSTEM_NAME STREQUAL Linux)
     if(ENABLE_NAT)
         add_definitions(-DENABLE_NAT)
     endif()
+
+    option(ENABLE_REUSE_PORT "Build with REUSE_PORT support" ON)
+    if (ENABLE_REUSE_PORT)
```
Comment (GreaterFire) [inline CMakeLists.txt:81]:
```suggestion
    if(ENABLE_REUSE_PORT)
```

**Review by GreaterFire — APPROVED** [review]
LGTM! Thanks!

**Pull request merged**